### PR TITLE
Get preview images on the fly and keep them in cache

### PR DESCRIPTION
**Pull request by azhavoro**
### Motivation and context

Improved image preview loading for **Tasks**, **Jobs** and **Projects** views
Backend behaviour change: creating image previews by request and storing them in the cache
Added corresponding endpoints:
tasks/{id}/preview
projects/{id}/preview
jobs/{id}/preview

Demonstration(added random 0-1s delay for demo purposes):

https://user-images.githubusercontent.com/41117609/208106321-951b8647-6e6b-452e-910c-31c4d0b8682d.mp4

https://user-images.githubusercontent.com/41117609/208106339-2d3a5a7b-d422-4b27-9e76-08729022e1ca.mp4

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have added a description of my changes into [CHANGELOG](https://github.com/cvat-ai/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/cvat-ai/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

**Review by nmanovic — APPROVED** [review]
From the fist looks the server part looks good! Let me test the PR.

**Comment by nmanovic** [comment]
@azhavoro , some tests are failed

```
ndex created!
[2254](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2255)
  File "/home/django/cvat/apps/engine/tests/test_rest_api.py", line 3370, in _test_api_v2_tasks_id_data_spec
[2255](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2256)
The following import formats have no pair: {'CVAT 1.1'}
[2256](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2257)
    self.assertLessEqual(preview.size, image_sizes[0])
[2257](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2258)
The following export formats have no pair: {'CVAT for video 1.1', 'CVAT for images 1.1'}
[2258](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2259)
AssertionError: (256, 52) not less than or equal to (244, 507)
[2259](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2260)
loading annotations into memory...
[2260](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2261)

[2261](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2262)
Done (t=0.00s)
[2262](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2263)
----------------------------------------------------------------------
[2263](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2264)
creating index...
[2264](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2265)
Ran 272 tests in 467.384s
[2265](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2266)
index created!
[2266](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2267)

[2267](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2268)
The following import formats have no pair: {'CVAT 1.1'}
[2268](https://github.com/opencv/cvat/actions/runs/3713037113/jobs/6295588528#step:7:2269)
FAILED (failures=1, errors=2, skipped=36)
```

File: cvat-ui/src/reducers/jobs-reducer.ts [hunk 1]
```diff
@@ -45,6 +44,46 @@ export default (state: JobsState = defaultState, action: JobsActions): JobsState
...
+            const { previews } = state;
+            previews[jobID] = {
+                preview: '',
+                fetching: true,
+                initialized: false,
+            };
+            return {
+                ...state,
+                previews,
+            };
```
Comment (bsekachev) [inline cvat-ui/src/reducers/jobs-reducer.ts:58]: Better to make this code desructive following functional concepts, like:

```js
return {
    ...state,
    previews: {
        ...previews,
        [jobID]: {
                preview: '',
                fetching: true,
                initialized: false,
        },
    },
}
```

File: cvat-ui/src/reducers/jobs-reducer.ts [hunk 2]
```diff
@@ -45,6 +44,46 @@ export default (state: JobsState = defaultState, action: JobsActions): JobsState
...
+        }
+        case JobsActionTypes.GET_JOB_PREVIEW_SUCCESS: {
+            const { jobID, preview } = action.payload;
+            const { previews } = state;
+            previews[jobID] = {
+                ...previews[jobID],
+                preview,
+                initialized: true,
+                fetching: false,
+            };
```
Comment (bsekachev) [inline cvat-ui/src/reducers/jobs-reducer.ts:68]: The same comment as above and the same about failed actions and other reducers

**Comment by bsekachev** [comment]
Generally, fine. But could you please look license headers and reduce code duplication according to proposed ideas?

File: cvat/apps/engine/cache.py [hunk 3]
```diff
@@ -25,15 +25,24 @@ def __init__(self, dimension=DimensionType.DIM_2D):
...
         if not chunk:
-            chunk, tag = self.prepare_chunk_buff(db_data, quality, chunk_number)
-            self.save_chunk(db_data.id, chunk_number, quality, chunk, tag)
+            chunk, tag = self._prepare_chunk_buff(db_data, quality, chunk_number)
+            self._cache.set(cache_key, chunk, tag=tag)
+
         return chunk, tag
 
-    def prepare_chunk_buff(self, db_data, quality, chunk_number):
+    def get_preview(self, key):
```
Comment (nmanovic) [inline cvat/apps/engine/cache.py:38]: @azhavoro , may be it is a problem. For get_chunk_mime the cache decide how to name the key. For get_preview, it is decided outside.

Should we fix that?

File: cvat/apps/engine/migrations/0062_delete_previews.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from django.db import migrations
+from django.conf import settings
+
+def delete_previews(apps, schema_editor):
+    migration_name = os.path.splitext(os.path.basename(__file__))[0]
+    migration_log_file = '{}.log'.format(migration_name)
+    stdout = sys.stdout
+    stderr = sys.stderr
+    # redirect all stdout to the file
+    log_file_object = open(os.path.join(settings.MIGRATIONS_LOGS_ROOT, migration_log_file), 'w')
```
Comment (nmanovic) [inline cvat/apps/engine/migrations/0062_delete_previews.py:15]: I think, that the logging capabilities are necessary for all other migration files as well. I see it is the second while where we initialize logging.

0024_auto_20191023_1025.py

File: cvat/apps/engine/migrations/0062_delete_previews.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                log.error(str(e))
+                traceback.print_exc(file=sys.stderr)
+
+    log.info('\nDeleting Data previews...')
+    Data = apps.get_model('engine', 'Data')
+    delete_object_previews(Data.objects.all(), settings.MEDIA_DATA_ROOT)
+
+    log.info('\nDeleting Job previews...')
+    Job = apps.get_model('engine', 'Job')
+    delete_object_previews(Job.objects.all(), settings.JOBS_ROOT)
```
Comment (nmanovic) [inline cvat/apps/engine/migrations/0062_delete_previews.py:40]: Potentially we can have millions of jobs. The line can be a problem from the performance point of view. On one side it is the right approach. On another side, if we just remove all `{settings.JOBS_ROOT}/{num}/preview.jpeg` files will it be a problem?

**Comment by nmanovic** [comment]
@azhavoro , any updates?

File: cvat/apps/engine/migrations/0062_delete_previews.py (hunk 5 above)
Comment (azhavoro) [reply cvat/apps/engine/migrations/0062_delete_previews.py]: >>> def f():
...   now =datetime.now()
...   l = list(Job.objects.all())
...   print(datetime.now() - now)
...   print(len(l))
...
>>> f()
0:00:09.785674
1015453

File: cvat-ui/src/reducers/index.ts [hunk 6]
```diff
@@ -90,12 +97,28 @@ export interface JobsQuery {
...
+    current: Job[];
+    previews: {
+        [index: number]: JobPreview;
+    };
+}
+
+interface TaskPreview {
+    fetching: boolean;
+    initialized: boolean;
+    preview: string;
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/reducers/index.ts:117]: Still have the same question :)

File: cvat/apps/engine/migrations/0062_delete_previews.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+from django.db import migrations
+from django.conf import settings
+from cvat.apps.engine.log import get_migration_logger
+
+def delete_previews(apps, schema_editor):
+    migration_name = os.path.splitext(os.path.basename(__file__))[0]
+    with get_migration_logger(migration_name) as log:
+        def delete_object_previews(db_objects, root_path):
+            for db_obj in db_objects:
```
Comment (nmanovic) [inline cvat/apps/engine/migrations/0062_delete_previews.py:13]: Do we really need to work with DB objects here? Will it be better and faster just remove all root_path/**/preview.jpeg? What is your opinion?

File: cvat/apps/engine/migrations/0062_delete_previews.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+                    log.error(str(e))
+                    traceback.print_exc(file=sys.stderr)
+
+        log.info('\nDeleting Data previews...')
+        Data = apps.get_model('engine', 'Data')
+        delete_object_previews(Data.objects.all(), settings.MEDIA_DATA_ROOT)
+
+        log.info('\nDeleting Job previews...')
+        Job = apps.get_model('engine', 'Job')
+        delete_object_previews(Job.objects.all(), settings.JOBS_ROOT)
```
Comment (nmanovic) [inline cvat/apps/engine/migrations/0062_delete_previews.py:28]: It can be a potentially a huge request.

**Review by bsekachev — APPROVED** [review]

File: cvat/apps/engine/views.py [hunk 9]
```diff
@@ -622,6 +621,28 @@ def import_backup(self, request, pk=None):
     def append_backup_chunk(self, request, file_id):
         return self.append_tus_chunk(request, file_id)
 
+    @extend_schema(summary='Method returns a preview image for the project',
+        responses={
+            '200': OpenApiResponse(description='Project image preview'),
```
Comment (nmanovic) [inline cvat/apps/engine/views.py:626]: It can return 404 as I can see. Can it return anything else?

File: cvat/apps/engine/views.py [hunk 10]
```diff
@@ -1317,6 +1339,24 @@ def dataset_export(self, request, pk):
             return Response(data="Exporting a dataset from a task without data is not allowed",
                 status=status.HTTP_400_BAD_REQUEST)
 
+    @extend_schema(summary='Method returns a preview image for the task',
+        responses={
+            '200': OpenApiResponse(description='Task image preview'),
```
Comment (nmanovic) [inline cvat/apps/engine/views.py:1344]: @azhavoro , what it will return if a task does not have any images?

Comment (azhavoro) [reply cvat/apps/engine/views.py:1346]: fixed

File: cvat/apps/engine/migrations/0062_delete_previews.py (hunk 8 above)
Comment (azhavoro) [reply cvat/apps/engine/migrations/0062_delete_previews.py:28]: Results of my experiment:
```python
def f():
...   count = 0
...   now = datetime.now()
...   for ff in glob.glob('/home/django/data/jobs/*/preview.jpeg'):
...     count += 1
...   print(datetime.now() - now)
...   print(count)
...
>>> f()
0:03:49.090189
53651
```
vs
```python
def f():
... now =datetime.now()
... l = list(Job.objects.all())
... print(datetime.now() - now)
... print(len(l))
...
f()
0:00:09.785674
1015453
```
Both approaches require `os.remove(file_path)` for each iteration

**Pull request merged**